ui/revenuecatui/src/main/kotlin/com/revenuecat/purchases/ui/revenuecatui/data/PaywallViewModel.kt

```diff
@@ -145,47 +157,79 @@ internal class PaywallViewModelImpl(
             Logger.e("Activity is null, not initiating package purchase")
             return
         }
-        when (val currentState = _state.value) {
-            is PaywallState.Loaded -> {
-                val selectedPackage = currentState.selectedPackage.value
-                if (!selectedPackage.currentlySubscribed) {
-                    purchasePackage(activity, selectedPackage.rcPackage)
-                } else {
-                    Logger.d("Ignoring purchase request for already subscribed package")
-                }
-            }
+        viewModelScope.launch {
+            handlePackagePurchase(activity)
+        }
+    }
 
-            else -> {
-                Logger.e("Unexpected state trying to purchase package: $currentState")
-            }
+    override suspend fun awaitPurchaseSelectedPackage(activity: Activity?) {
+        if (activity == null) {
+            Logger.e("Activity is null, not initiating package purchase")
+            return
         }
+        handlePackagePurchase(activity)
     }
 
-    override fun restorePurchases() {
+    @Suppress("NestedBlockDepth")
+    override suspend fun awaitRestorePurchases() {
         if (verifyNoActionInProgressOrStartAction()) { return }
+        try {
+            listener?.onRestoreStarted()
 
-        viewModelScope.launch {
-            try {
-                listener?.onRestoreStarted()
+            val customRestoreHandler = myAppPurchaseLogic?.performRestore
 
-                val customerInfo = purchases.awaitRestore()
-
-                Logger.i("Restore purchases successful: $customerInfo")
-                listener?.onRestoreCompleted(customerInfo)
-
-                shouldDisplayBlock?.let {
-                    if (!it(customerInfo)) {
-                        Logger.d("Dismissing paywall after restore since display condition has not been met")
-                        options.dismissRequest()
+            when (purchases.purchasesAreCompletedBy) {
+                PurchasesAreCompletedBy.MY_APP -> {
+                    checkNotNull(customRestoreHandler) {
+                        "myAppPurchaseLogic must not be null when purchases.purchasesAreCompletedBy " +
+                            "is PurchasesAreCompletedBy.MY_APP"
+                    }
+                    val customerInfo = purchases.awaitCustomerInfo()
+                    when (val result = customRestoreHandler.invoke(customerInfo)) {
```

Comment:
`invoke` can be omitted: 
 ```suggestion
                     when (val result = customRestoreHandler(customerInfo)) {
```

Reply:
Done!